Add key_padding_mask kwarg to Transformer

lucasgadams opened the pull request:
Motivation:
The forward method of MultiheadAttention has a kwarg a key_padding_mask. This mask is of shape (N,S) where N is batch and S is sequence length. This mask is applied prior to attention softmax where True values in the mask are set to float('-inf'). This allows you to mask position j from attention for all position i in input sequence. It's typically used to mask padded inputs. So for a sample in a batch we will be able to make sure no encoder outputs depend on padding inputs. Currently the Transformer, TransformerEncoder, and TransformerEncoderLayer do not have this kwarg, and only have options for a (S,S), (T,T), and (S,T) masks which are applied equally across the batch for source input, target output, and target-source memory respectively. These masks can't be used for padding and are instead used for things like subsequent masking in language modeling, by masking the attention of position i to position j. 

This diff exposes the key_padding_mask to Transformer, TransformerEncoder, and TransformerEncoderLayer forward methods which is ultimately passed to MultiheadAttention forward.

Open question: should we also allow a key_padding_mask for the decoder layer? As padding is usually at the end of each sentence in a batch and sentences are usually decoding from left to right, usually people deal with padding on decoded outputs by just masking those outputs at the loss layer. There might be some scenarios where it's needed though I don't think it would be common. People can also still just subclass and override the layers. We could also pass the input key_padding_mask to the memory <> decoder attention layer. Not sure if that's necessary though because the output of position i from each attention encoder layer won't depend on any masked positions in the input (even if position i is a masked position itself) so there's not really any point in masking position i again.



Summary:
Adds the key_padding_mask kwarg to Transformer, TransformerEncoder, and TransformerEncoderLayer forward methods.
The standard TransformerEncoderLayer uses a MultiheadAttention layer as self_attn. MultiheadAttention forward method has a key_padding_mask kwarg that allows for masking of values such as padding per sequence in a batch, in contrast to the attn_mask kwarg which is usually of shape (S,S) and applied equally across the batch.

MultiheadAttention calls functional.multi_head_attention_forward, which has the same key_padding_mask kwarg of shape (N,S). Masked (True) values are set to float('-inf').

Differential Revision: D16112263

zhangguanheng66 commented:
fix https://github.com/pytorch/pytorch/issues/22374#issuecomment-507464094

zhangguanheng66 commented:
@sebamenabar I think this PR is relevant to your feature request. Feel free to add review comments. Thanks.

zhangguanheng66 commented:
@stephenroller  @myleott @ngimel @DNGros @mttk a PR to add key_padding to transformer_encoder. Just wondering if we should do the same thing for transformer_decoder.

zhangguanheng66 commented:
@lucasgadams please update the text with motivation for the PR.

zhangguanheng66 commented on test/test_nn.py:
```diff
@@ -5801,8 +5801,10 @@ def test_Transformer_cell(self):
+                             memory_mask=memory_mask, key_padding_mask=key_padding_mask)
         output.sum().backward()
 
     def test_transformerencoderlayer(self):
```
Though we didn't do the deterministic test here for the kwarg, it would be great if you could add the deterministic test for key_padding_mask.

lucasgadams replied:
Added

zhangguanheng66 reviewed (changes requested):
After some discussions, I feel it makes some sense to have the key_padding in the decoder (at least for the self-attention layer). We prefer to maintain a generic API because we can't predict application on the user side.

sebamenabar commented:
> After some discussions, I feel it makes some sense to have the key_padding in the decoder (at least for the self-attention layer). We prefer to maintain a generic API because we can't predict application on the user side.

Agree, I think it also makes sense to add the key_padding_mask on the decoder's multihead attention that looks at the encoder output, since we usually have padding on the source and the target.

lucasgadams commented:
Ok I have added the key_padding_mask kwarg to the Decoder and DecoderLayer which will pass it in to  the target - memory attention. 1 more question is if we want to automatically pass the src key_padding_mask to both the encoder and decoder in Transformer forward method. This means people will be forced to used it on both if they use it on either in Transformer. I think this is probably fine and so that is what I have done in the latest diff. But let me know if you think we don't want to automatically do that in Transformer. I will add some test cases for the decoder piece.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -280,22 +296,25 @@ def __init__(self, d_model, nhead, dim_feedforward=2048, dropout=0.1):
         tgt = self.norm1(tgt)
-        tgt2 = self.multihead_attn(tgt, memory, memory, attn_mask=memory_mask)[0]
+        tgt2 = self.multihead_attn(tgt, memory, memory, attn_mask=memory_mask,
+                                   key_padding_mask=key_padding_mask)[0]
```
I think we want to add the key_padding_mask to the self-attention layer.

lucasgadams replied:
Not the memory layer? I would think people would want to mask the padding pieces on the input side when decoding and that would be the way to achieve it. I suppose you might want to mask padding on the target. How about I just add key_padding_mask to self attention layer and memory_key_padding_mask arg for memory layer. In Transformer I will add source_key_padding_mask, target_key_padding_mask, and memory_key_padding_mask. That will cover every base that we would need and be as flexible as possible.

zhangguanheng66 replied:
I agree this will cover every base enough.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -55,26 +55,35 @@ def __init__(self, d_model=512, nhead=8, num_encoder_layers=6,
 
-    def forward(self, src, tgt, src_mask=None, tgt_mask=None, memory_mask=None):
+    def forward(self, src, tgt, src_mask=None, tgt_mask=None,
+                memory_mask=None, key_padding_mask=None):
```
the key_padding masks for encoder and decoder are not necessary to be same, though they could be. So maybe it's better to have encoder_key_padding_mask and decoder_key_padding_mask and both are passed to the self-attention layer in encoder/decoder, respectively.

lucasgadams replied:
I think you still want the option of masking the memory layer, as @sebamenabar pointed out

zhangguanheng66 replied:
So I have no explicit opposition on this. :)

zhangguanheng66 commented:
> Ok I have added the key_padding_mask kwarg to the Decoder and DecoderLayer which will pass it in to the target - memory attention. 1 more question is if we want to automatically pass the src key_padding_mask to both the encoder and decoder in Transformer forward method. This means people will be forced to used it on both if they use it on either in Transformer. I think this is probably fine and so that is what I have done in the latest diff. But let me know if you think we don't want to automatically do that in Transformer. I will add some test cases for the decoder piece.

See my comments. I don't think we want to pass a single key_padding to both encoder and decoder. Thanks.

lucasgadams commented:
I added the requested changes. Transformer.forward now uses src_key_padding_mask, tgt_key_padding_mask, and memory_key_padding_mask. Decoder uses key_padding_mask and memory_key_padding_mask. Modified the relevant tests and added some new deterministic ones for the decoder masks.

sebamenabar commented:
> I added the requested changes. Transformer.forward now uses src_key_padding_mask, tgt_key_padding_mask, and memory_key_padding_mask. Decoder uses key_padding_mask and memory_key_padding_mask. Modified the relevant tests and added some new deterministic ones for the decoder masks.

I think every thing is covered now, arguments look very long, but in my opinion all options are standard use cases of the transformer.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -55,26 +55,40 @@ def __init__(self, d_model=512, nhead=8, num_encoder_layers=6,
+            that should be masked with float('-inf') and False values will be unchanged.
+            This mask ensures that no information will be taken from position i if
+            it is masked, and has a separate mask for each sequence in a batch. Only
+            applied to the encoder.
```
This is not necessarily true now.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -174,22 +191,28 @@ def __init__(self, decoder_layer, num_layers, norm=None):
-            tgt_mask: the mask for the tgt sequence (optional).
+            mask: the mask for the tgt sequence (optional).
             memory_mask: the mask for the memory sequence (optional).
+            key_padding_mask: the mask for the tgt keys per batch (optional).
```
For consistency with Transformer, I think we can use "tgt_key_padding_mask" here, instead of key_padding_mask. Sorry for being annoying.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -134,20 +150,21 @@ def __init__(self, encoder_layer, num_layers, norm=None):
             src: the sequnce to the encoder (required).
-            src_mask: the mask for the src sequence (optional).
+            mask: the mask for the src sequence (optional).
+            key_padding_mask: the mask for the src keys per batch (optional).
```
For consistency with Transformer, I think we can use "src_key_padding_mask" here, instead of key_padding_mask.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -174,22 +191,28 @@ def __init__(self, decoder_layer, num_layers, norm=None):
             tgt: the sequence to the decoder (required).
             memory: the sequnce from the last layer of the encoder (required).
-            tgt_mask: the mask for the tgt sequence (optional).
+            mask: the mask for the tgt sequence (optional).
```
why change to mask, instead of "tgt_mask"? Change it will break the usage of our current users since they have to switch to the new APIs. Unless there is a solid reason, I prefer not to do so.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -227,17 +250,19 @@ def __init__(self, d_model, nhead, dim_feedforward=2048, dropout=0.1):
         self.dropout2 = Dropout(dropout)
 
-    def forward(self, src, src_mask=None):
+    def forward(self, src, mask=None, key_padding_mask=None):
```
Similarly, no changes to the APIs. Backward capability issue.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -280,22 +305,27 @@ def __init__(self, d_model, nhead, dim_feedforward=2048, dropout=0.1):
         self.dropout3 = Dropout(dropout)
 
-    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
+    def forward(self, tgt, memory, mask=None, memory_mask=None,
```
Similarly, no changes to the APIs. Backward capability issue.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -174,22 +191,28 @@ def __init__(self, decoder_layer, num_layers, norm=None):
         self.norm = norm
 
-    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
+    def forward(self, tgt, memory, mask=None,
```
Similarly, no changes to the APIs. Backward capability issue. See the comment below.

zhangguanheng66 reviewed (changes requested):
.

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -55,26 +55,39 @@ def __init__(self, d_model=512, nhead=8, num_encoder_layers=6,
-            for position i depend only on the information before position i.
+            - src_key_padding_mask: :math:`(N, S)`.
+            - tgt_key_padding_mask: :math:`(N, T)`.
+            - memory_key_padding_mask: :math:`(N, S)`.
```
Since memory_key_padding_mask is used on the multihead_attention layer in decoder, it's dimension should be consistent with that of tgt, rather than src. I guess it is just a typo. 
I'm wondering if we should really have a memory_key_padding_mask, unless a standalone TransformerDecoder is used and memory is passed to the decoder. In that situation, I think the answer is yes, right?

lucasgadams replied:
I don't think that is True. Memory_key_padding_mask is used on the memory, which is the same dimensions as src. You can see in my tests I always use src length for the memory mask.

lucasgadams replied:
In regard to your second question, memory_key_padding_mask give the user the flexibility to not attend to positions in the memory that were padded on the src input.

zhangguanheng66 commented:
@pytorchbot rebase and test

lucasgadams commented:
@pytorchbot rebase this please

lucasgadams committed:
Add key_padding_mask kwarg to Transformer (#22588)

Summary:
Motivation:
The forward method of MultiheadAttention has a kwarg a key_padding_mask. This mask is of shape (N,S) where N is batch and S is sequence length. This mask is applied prior to attention softmax where True values in the mask are set to float('-inf'). This allows you to mask position j from attention for all position i in input sequence. It's typically used to mask padded inputs. So for a sample in a batch we will be able to make sure no encoder outputs depend on padding inputs. Currently the Transformer, TransformerEncoder, and TransformerEncoderLayer do not have this kwarg, and only have options for a (S,S), (T,T), and (S,T) masks which are applied equally across the batch for source input, target output, and target-source memory respectively. These masks can't be used for padding and are instead used for things like subsequent masking in language modeling, by masking the attention of position i to position j.

This diff exposes the key_padding_mask to Transformer, TransformerEncoder, and TransformerEncoderLayer forward methods which is ultimately passed to MultiheadAttention forward.

Open question: should we also allow a key_padding_mask for the decoder layer? As padding is usually at the end of each sentence in a batch and sentences are usually decoding from left to right, usually people deal with padding on decoded outputs by just masking those outputs at the loss layer. There might be some scenarios where it's needed though I don't think it would be common. People can also still just subclass and override the layers. We could also pass the input key_padding_mask to the memory <> decoder attention layer. Not sure if that's necessary though because the output of position i from each attention encoder layer won't depend on any masked positions in the input (even if position i is a masked position itself) so there's not really any point in masking position i again.
Adds the key_padding_mask kwarg to Transformer, TransformerEncoder, and TransformerEncoderLayer forward methods.
The standard TransformerEncoderLayer uses a MultiheadAttention layer as self_attn. MultiheadAttention forward method has a key_padding_mask kwarg that allows for masking of values such as padding per sequence in a batch, in contrast to the attn_mask kwarg which is usually of shape (S,S) and applied equally across the batch.

MultiheadAttention calls functional.multi_head_attention_forward, which has the same key_padding_mask kwarg of shape (N,S). Masked (True) values are set to float('-inf').
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22588

Test Plan:
buck test mode/dev caffe2/test:nn -- 'test_transformerencoderlayer \(test_nn\.TestNN\)'
buck test mode/dev caffe2/test:nn -- 'test_Transformer_cell \(test_nn\.TestNN\)'
buck test mode/dev caffe2/test:nn -- 'test_transformer_args_check \(test_nn\.TestNN\)'

Differential Revision: D16112263

fbshipit-source-id: c56e9ff409f6666253cfc9b1d23656981e6729d1
